packages/core/contracts/cross-chain-oracle/chain-adapters/Optimism_ParentMessenger.sol

```diff
@@ -0,0 +1,67 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+pragma solidity ^0.8.0;
+
+// This should be replaced with a "real" import when Optimism release their new contract versions.
+import "../../external/ovm/OVM_CrossDomainEnabled.sol";
+import "../interfaces/ParentMessengerInterface.sol";
+import "../interfaces/ParentMessengerConsumerInterface.sol";
+import "./ParentMessengerBase.sol";
+
+/**
+ * @notice Sends cross chain messages from Ethereum L1 to Optimism L2 network.
+ * @dev This contract's is ownable and should be owned by the DVM governor.
+ */
+contract Optimism_ParentMessenger is OVM_CrossDomainEnabled, ParentMessengerInterface, ParentMessengerBase {
+    event SetDefaultGasLimit(uint32 newDefaultGasLimit);
+    event MessageSentToChild(bytes data, address indexed childAddress, uint32 defaultGasLimit);
+    event MessageReceivedFromChild(bytes data, address indexed childAddress);
+
+    uint32 public defaultGasLimit = 5_000_000;
+
+    /**
+     * @notice Construct the Optimism_ParentMessenger contract.
+     * @param _crossDomainMessenger The address of the Optimism cross domain messenger contract.
+     * @param _childChainId The chain id of the Optimism L2 network this messenger should connect to.
+     **/
+    constructor(address _crossDomainMessenger, uint256 _childChainId)
+        OVM_CrossDomainEnabled(_crossDomainMessenger)
+        ParentMessengerBase(_childChainId)
+    {}
+
+    /**
+     * @notice Changes the default gas limit that is sent along with transactions to Optimism.
+     * @dev The caller of this function must be the owner. This should be set to the DVM governor.
+     * @param newDefaultGasLimit the new L2 gas limit to be set.
+     */
+    function setDefaultGasLimit(uint32 newDefaultGasLimit) public onlyOwner {
+        defaultGasLimit = newDefaultGasLimit;
+        emit SetDefaultGasLimit(newDefaultGasLimit);
+    }
+
+    /**
+     * @notice Sends a message to the child messenger via the canonical message bridge.
+     * @dev The caller must be the either the OracleHub or the GovernorHub, deployed on L2. This is to send either a
+     * price or initiate a governance action on L2.
+     * @dev The recipient of this message is the child messenger. The messenger must implement processMessageFromParent
+     * which then forwards the data to the target either the OracleSpoke or the governorSpoke depending on the caller.
+     * @param data data message sent to the L2 messenger. Should be an encoded function call or packed data.
+     */
+    function sendMessageToChild(bytes memory data) public override onlyHubContract() {
+        address target = msg.sender == oracleHub ? oracleSpoke : governorSpoke;
+        bytes memory dataSentToChild = abi.encodeWithSignature("processMessageFromParent(bytes,address)", data, target);
```

Comment:
As discussed offline, I think the chains of contracts that these messages have to pass through:
1. Makes them a bit more error prone as there is more configuration surface area.
2. Makes them more confusing to trace/debug.
3. Creates more opportunities for spoofing where a permissioning code error allows another contract to pretend to be one of our mainnet contracts.

Have we considered shortening the chain by putting these in the _inheritance tree_ for the hub and spoke governor/oracle contracts?

This might look something like:
```
contract OptimismGovernorSpoke is GovernorSpoke, OptimismChildMessenger {
...
}
```

Then, consuming these messages would just be calling internal methods, and we would only need to make sure that the caller on the other side was the GovernorHub.

This would introduce a problem that on the Hub side, you would need to deploy a separate Oracle and Governor hub for each chain rather than just storing a mapping to the messengers.

Not strongly opinionated as I think there isn't a perfect solution, but I just wanted to bring it up since the long chains of contracts makes me a bit nervous because of the overall surface area and configuration complexity.

Thoughts?

Reply:
> This would introduce a problem that on the Hub side, you would need to deploy a separate Oracle and Governor hub for each chain rather than just storing a mapping to the messengers.

Yeah this is my main push back. I think this proposed architecture works better for the Governor Hub/Spoke which doesn't have to store additional price request information, but deploying a separate OracleHub for each L2 is not a great solution. Thoughts?

Reply:
Agreed with @nicholaspai . the whole point of this design is that we have one OracleHub/GovernerHub for n number of chains. I agree that in the simple 1:1 case (only L1) the pattern here is over-engineered and complex but if we onboard 5 chains do you think having it as part of inheritance will be easier to reason about and manage? I'm not too sure.